jax/_src/api.py

```diff
@@ -189,40 +191,35 @@ def _isgeneratorfunction(fun):
 _POSITIONAL_OR_KEYWORD = inspect.Parameter.POSITIONAL_OR_KEYWORD
 
 def _infer_argnums_and_argnames(
-    fun: Callable,
+    sig: inspect.Signature,
     argnums: Union[int, Iterable[int], None],
     argnames: Union[str, Iterable[str], None],
   ) -> Tuple[Tuple[int, ...], Tuple[str, ...]]:
   """Infer missing argnums and argnames for a function with inspect."""
   if argnums is None and argnames is None:
-    argnums = ()
-    argnames = ()
-  elif argnums is not None and argnames is not None:
+    return (), ()
+
+  if argnums is not None and argnames is not None:
     argnums = _ensure_index_tuple(argnums)
     argnames = _ensure_str_tuple(argnames)
+
+    return argnums, argnames
+
+  parameters = sig.parameters
+  if argnums is None:
+    assert argnames is not None
```

Comment:
Nit: I think the assert is redundant at this point

Reply:
The assert is to aid `mypy`, as we get an error otherwise. Alternatively one could use a `typing.cast`, but I prefer the assert.

Further, this was already in place prior to this PR.